MAVProxy/modules/mavproxy_speech.py

```diff
@@ -1,26 +1,209 @@
+            self.scale,
+            self.formatstr)
+
+class ConditionalSpeechAnnounce(PeriodicSpeechAnnounce):
+    '''
+    A SpeechAnnounce subclass that provides rate-limited announcements
+    only when a mavlink value is outside an expected range.
+    '''
+
+    def __init__(self, module, msg_type, entry_name, sayEveryS=5.0, sayAs=None, scale=1, formatstr="%0.0f", range=[-float("inf"), float("inf")]):
+        super(ConditionalSpeechAnnounce, self).__init__(module, msg_type, entry_name, sayEveryS=sayEveryS, sayAs=sayAs, scale=scale, formatstr=formatstr)
+        self.range = range
+
+    def set_min(self, minval):
+        self.range[1] = minval
+
+    def set_max(self, maxval):
+        self.range[1] = maxval
+
+    def maybe_say(self, msg, backend):
+        value = self.msg_entry_to_text(msg)
+        if value < self.range[0] or value > self.range[1]:
+            super(ConditionalSpeechAnnounce, self).maybe_say(msg, backend)
+
+    def to_speech_text(self, value):
+        if value < self.range[0]:
+            return "%s below minimum! Expected minimum %s, current value %s." % (
+                self.get_speech_name(), self.convert(self.range[0]), self.convert(value))
+        if value > self.range[1]:
+            return "%s above maximum! Expected maximum %s, current value %s." % (
+                self.get_speech_name(), self.convert(self.range[1]), self.convert(value))
+
+    def __str__(self):
+        s = super(ConditionalSpeechAnnounce, self).__str__()
+        return "%s when value is outside [%f, %f]" % (s, self.range[0], self.range[1])  
+
+class OnChangeSpeechAnnounce(PeriodicSpeechAnnounce):
+    '''
+    A SpeechAnnounce subclass that provides rate-limited announcements
+    only when a mavlink value changes.
+    '''
+
+    def __init__(self, module, msg_type, entry_name, sayEveryS=5.0, sayAs=None, scale=1, formatstr="%0.0f", threshold=0):
+        super(OnChangeSpeechAnnounce, self).__init__(module, msg_type, entry_name, sayEveryS=sayEveryS, sayAs=sayAs, scale=scale, formatstr=formatstr)
+        self.prev_value = float("inf")
+        self.threshold = threshold
+
+    def set_threshold(self, threshold):
+        self.threshold = threshold
+
+    def maybe_say(self, msg, backend):
+        new_value = self.msg_entry_to_text(msg)
+        if self.prev_value == float("inf"):
+            self.prev_value = new_value
+        if abs(self.prev_value - new_value) >= self.threshold:
+            super(OnChangeSpeechAnnounce, self).maybe_say(msg, backend)
+
+    def to_speech_text(self, value):
+        ret = "%s changing from %s to %s" % (self.get_speech_name(), self.convert(self.prev_value), self.convert(value))
+        self.prev_value = value
+        return ret
+
+    def __str__(self):
+        s = super(OnChangeSpeechAnnounce, self).__str__()
+        return "%s when value changes by more than %f" % (s, self.threshold)  
+
+import threading
+import Queue
+import subprocess
+
+class SaySpeechBackend(threading.Thread):
```

Comment:
This is a separate change to the "conditional announcement" stuff etc, right?

Could we get this in a separate patch so it might get merged earlier?